src/DGmethods/examples/ex_001_periodic_advection.jl

```diff
@@ -0,0 +1,438 @@
+# ```math
+# f^{*} = f^{*}(q^{-}, q^{+}; \vec{n}).
+# ```
+# We call the two sides of the interface the "minus side" and "plus side" hence
+# the $\pm$ superscripts in $q$ in the notation above; the choice of which
+# element is on the minus side and plus side is arbitrary. Here $\vec{n}$ is a
+# unit normal to the interface and is typically taken to point from the minus
+# side to the plus side. The numerical flux is required to be symmetric with
+# respect to $q^{-}$ and $q^{+}$:
+# ```math
+# f^{*}(q^{-}, q^{+}; \vec{n}) = f^{*}(q^{+}, q^{-}; \vec{n}).
+# ```
+# and consistent with the physical flux in the sense that
+# ```math
+# \boldsymbol{F}(q) \cdot \vec{n} = f^{*}(q, q; \vec{n}).
+# ```
+# Taken together these two conditions also imply that
+# ```math
+# f^{*}(q^{-}, q^{+}; \vec{n}^{-}) = -f^{*}(q^{+}, q^{-}; \vec{n}^{+}),
+# ```
+# that is $f^{*}$ is skew-symmetry with respect to the unit normals $\vec{n}^{+}
+# = -\vec{n}^{-}$.  The choice of numerical flux has important implications for
+# the stability of the method. Though it is beyond the scope of these tutorials
+# to dive into this in detail, often the "best" numerical fluxes are the ones
+# that are constructed specifically for given set of equations by solving the
+# Riemann problem; the Riemann problem is an initial value problem where the
+# initial data are piecewise constant with a single discontinuity.
+#
+# In the CLIMA balance law solver the numerical flux function is a user-defined
+# function that fills in an `MVector` for the numerical flux given two states,
+# a unit normal to the face, the simulation time, and a user-defined auxiliary
+# state; the auxiliary state will be discussed in a subsequent example. In the
+# function below `F` is the numerical flux to fill, `nM` is the unit normal
+# pointing away from the minus side and toward the plus side, `QM` and `QP` are
+# `MVector`s of the solution state on the minus  and plus sides of the interface,
+# similar `auxM` and `auxP` are the user-defined auxiliary state values on the
+# minus  and plus sides, and `t` is the simulation time.
+#
+# For linear advection the solution to the Riemann problem is trivial, since
+# if $\vec{n}^{-} \cdot \vec{u} ≥ 0$ the state $q$ is being advected from
+# the minus side to the plus side otherwise the reverse occurs. Thus the upwind
+# numerical flux for advection is
+# ```math
+# f^{*}(q^{-}, q^{+}; \vec{n}^{-}) =
+# \begin{cases}
+#   \vec{n}^{-}\cdot\vec{u} \; q^{-}, &\text{ if } \vec{n} \cdot \vec{u} ≥ 0,\\
+#   \vec{n}^{-}\cdot\vec{u} \; q^{+}, &\text{ otherwise.}
+# \end{cases}
+# ```
+# This is done in the following function
+function upwindflux!(fs, nM, stateM, auxM, stateP, auxP, t)
+  DFloat = eltype(fs)
+  @inbounds begin
+    ## determine the advection speed and direction
+    un = dot(nM, DFloat.(uvec))
+    qM = stateM[1]
+    qP = stateP[1]
+    ## Determine which state is "upwind" of the minus side
+    fs[1] = un ≥ 0 ? un * qM : un * qP
```

Comment:
I wonder if `fs[1] = elseif(un ≥ 0, un * qM, un * qP)` would be faster on a GPU.

Reply:
I would be surprised.

On the CPU they seem to generate the llvm same code (though I do not know what `spec.select` is):
 
```
julia> function max_ternary(a,b)
       a > b ? a : b
       end
max_ternary (generic function with 1 method)

julia> @code_llvm max_ternary(1,2)

;  @ REPL[30]:2 within `max_ternary'
define i64 @julia_max_ternary_12352(i64, i64) {
top:
; ┌ @ operators.jl:286 within `>'
; │┌ @ int.jl:49 within `<'
    %2 = icmp slt i64 %1, %0
; └└
  %spec.select = select i1 %2, i64 %0, i64 %1
  ret i64 %spec.select
}

julia> function max_ifelse(a,b)
       ifelse(a > b, a, b)
       end
max_ifelse (generic function with 1 method)

julia> @code_llvm max_ifelse(1,2)

;  @ REPL[32]:2 within `max_ifelse'
define i64 @julia_max_ifelse_12354(i64, i64) {
top:
; ┌ @ operators.jl:286 within `>'
; │┌ @ int.jl:49 within `<'
    %2 = icmp slt i64 %1, %0
; └└
  %3 = select i1 %2, i64 %0, i64 %1
  ret i64 %3
}
```

Reply:
Ah interesting. I think the way it was explained to me was that `ifelse` eagerly evaluates both branches so the GPU doesn't have to branch which might speed things up if the expressions is evaluated many times. But I must confess that I've never really benchmarked the difference.

```
help?> ifelse
search: ifelse

  ifelse(condition::Bool, x, y)

  Return x if condition is true, otherwise return y. This differs from ? or if in that it is an ordinary function, so all the arguments are evaluated first. In some cases, using ifelse instead of
  an if statement can eliminate the branch in generated code and provide higher performance in tight loops.

  Examples
  ≡≡≡≡≡≡≡≡≡≡

  julia> ifelse(1 > 2, 1, 2)
  2
```

Reply:
Interesting. I tried to dig into the julia source to see how there were implemented, but didn't get very far. Might be a question for @vchuravy